File: backend/app/api/schemas/plan_check.py
# -*- coding: utf-8 -*-
"""对照计划检查的响应。"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

PlanOrigin = Literal["synopsis", "beat"]
PlanBasis = Literal["literal", "model"]
PlanFreshness = Literal["unchecked", "current", "stale"]
PlanSource = Literal["model", "literal", "empty"]
PlanOutcome = Literal["unchecked", "no_plan", "gaps", "partial", "clear"]
BeatKind = Literal["plant", "advance", "payoff"]


class PlanCheckGap(BaseModel):
    """一条对得上梗概句子或本章节拍的缺口。"""

    ref: str = Field(description="计划条目编号")
    origin: PlanOrigin = Field(description="来自梗概或本章节拍")
    plan_text: str = Field(description="梗概原句，或这一条节拍的备注和意图")
    basis: PlanBasis = Field(description="模型判断或字面未出现")
    missing: list[str] = Field(description="正文里完全没有的专名或短句")
    detail: str = Field(description="模型说明；字面检查为空")
    beat_kind: BeatKind | None = Field(description="节拍类型")
    thread_name: str | None = Field(description="情节线名称")


class PlanCheckLine(BaseModel):
    """字面检查无法判断的一条计划。"""

    ref: str
    origin: PlanOrigin
    plan_text: str
    beat_kind: BeatKind | None = None
    thread_name: str | None = None


class PlanCheckResponse(BaseModel):
    """留在这一章上的对照结果。freshness=stale 表示正文或计划已经变了。"""

    chapter_id: str
    has_plan: bool = Field(description="当前章有梗概或本章节拍")
    freshness: PlanFreshness
    source: PlanSource | None = None
    outcome: PlanOutcome
    gaps: list[PlanCheckGap]
    unchecked: list[PlanCheckLine]
    checked_at: datetime | None = None

File: frontend/src/lib/plan-check.ts
import type { PlotBeatKind } from "./plot-thread";
import { normalizePlotBeatKind } from "./plot-thread";

export type PlanCheckOrigin = "synopsis" | "beat";
export type PlanCheckBasis = "literal" | "model";
export type PlanCheckFreshness = "unchecked" | "current" | "stale";
export type PlanCheckSource = "model" | "literal" | "empty";
export type PlanCheckOutcome = "unchecked" | "no_plan" | "gaps" | "partial" | "clear";

export interface PlanCheckGap {
  ref: string;
  origin: PlanCheckOrigin;
  planText: string;
  basis: PlanCheckBasis;
  missing: string[];
  detail: string;
  beatKind: PlotBeatKind | null;
  threadName: string | null;
}

export interface PlanCheckLine {
  ref: string;
  origin: PlanCheckOrigin;
  planText: string;
  beatKind: PlotBeatKind | null;
  threadName: string | null;
}

export interface PlanCheck {
  chapterId: string;
  hasPlan: boolean;
  freshness: PlanCheckFreshness;
  source: PlanCheckSource | null;
  outcome: PlanCheckOutcome;
  gaps: PlanCheckGap[];
  unchecked: PlanCheckLine[];
  checkedAt: string | null;
}

const FRESHNESS = new Set<PlanCheckFreshness>(["unchecked", "current", "stale"]);
const SOURCES = new Set<PlanCheckSource>(["model", "literal", "empty"]);
const OUTCOMES = new Set<PlanCheckOutcome>(["unchecked", "no_plan", "gaps", "partial", "clear"]);

function readOrigin(value: unknown): PlanCheckOrigin {
  return value === "beat" ? "beat" : "synopsis";
}

function readBasis(value: unknown): PlanCheckBasis {
  return value === "model" ? "model" : "literal";
}

function readStringList(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((item): item is string => typeof item === "string" && item.length > 0);
}

function readBeatKind(value: unknown): PlotBeatKind | null {
  if (typeof value !== "string") return null;
  return normalizePlotBeatKind(value);
}

export function transformPlanCheck(raw: Record<string, unknown>): PlanCheck {
  const gaps = Array.isArray(raw.gaps) ? raw.gaps : [];
  const unchecked = Array.isArray(raw.unchecked) ? raw.unchecked : [];
  const freshness = FRESHNESS.has(raw.freshness as PlanCheckFreshness)
    ? (raw.freshness as PlanCheckFreshness)
    : "unchecked";
  const source = SOURCES.has(raw.source as PlanCheckSource)
    ? (raw.source as PlanCheckSource)
    : null;
  const outcome = OUTCOMES.has(raw.outcome as PlanCheckOutcome)
    ? (raw.outcome as PlanCheckOutcome)
    : "unchecked";
  return {
    chapterId: typeof raw.chapter_id === "string" ? raw.chapter_id : "",
    hasPlan: raw.has_plan === true,
    freshness,
    source,
    outcome,
    gaps: gaps
      .filter((item): item is Record<string, unknown> => !!item && typeof item === "object")
      .map((item) => ({
        ref: typeof item.ref === "string" ? item.ref : "",
        origin: readOrigin(item.origin),
        planText: typeof item.plan_text === "string" ? item.plan_text : "",
        basis: readBasis(item.basis),
        missing: readStringList(item.missing),
        detail: typeof item.detail === "string" ? item.detail : "",
        beatKind: readBeatKind(item.beat_kind),
        threadName: typeof item.thread_name === "string" ? item.thread_name : null,
      })),
    unchecked: unchecked
      .filter((item): item is Record<string, unknown> => !!item && typeof item === "object")
      .map((item) => ({
        ref: typeof item.ref === "string" ? item.ref : "",
        origin: readOrigin(item.origin),
        planText: typeof item.plan_text === "string" ? item.plan_text : "",
        beatKind: readBeatKind(item.beat_kind),
        threadName: typeof item.thread_name === "string" ? item.thread_name : null,
      })),
    checkedAt: typeof raw.checked_at === "string" ? raw.checked_at : null,
  };
}
